File: app/schemas/role.py
"""
角色Schema
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class RoleBase(BaseModel):
    """角色基础Schema"""
    name: str = Field(..., min_length=1, max_length=100, description="角色名称")
    description: Optional[str] = Field(None, max_length=500, description="角色描述")
    tenant_id: Optional[str] = Field(None, description="租户ID，None表示系统级角色（如租户管理员）")
    status: bool = Field(True, description="状态：True-启用，False-停用")


class RoleCreate(RoleBase):
    """创建角色Schema"""
    pass


class RoleUpdate(BaseModel):
    """更新角色Schema"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="角色名称")
    description: Optional[str] = Field(None, max_length=500, description="角色描述")
    status: Optional[bool] = Field(None, description="状态：True-启用，False-停用")


class RoleResponse(RoleBase):
    """角色响应Schema"""
    id: str
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    
    class Config:
        from_attributes = True


class RoleWithPermissions(RoleResponse):
    """带权限的角色响应Schema"""
    permission_ids: List[str] = Field(default_factory=list, description="权限ID列表")
    
    class Config:
        from_attributes = True


class AssignPermissionsRequest(BaseModel):
    """分配权限请求Schema"""
    permission_ids: List[str] = Field(..., description="权限ID列表")


class AssignRolesRequest(BaseModel):
    """分配角色请求Schema"""
    role_ids: List[str] = Field(..., description="角色ID列表")


class AssignUsersRequest(BaseModel):
    """分配用户请求Schema"""
    user_ids: List[str] = Field(..., description="用户ID列表")

File: frontend/src/api/role.ts
import client from './client'

export interface Role {
  id: string
  name: string
  description?: string
  tenant_id: string | null  // 系统级角色为null
  status: boolean
  created_at: string
  updated_at: string
  created_by?: string
}

export interface RoleCreate {
  name: string
  description?: string
  tenant_id: string | null  // 系统级角色为null
  status?: boolean
}

export interface RoleUpdate {
  name?: string
  description?: string
  status?: boolean
}

export interface RoleListParams {
  skip?: number
  limit?: number
  status?: boolean
}

export interface AssignPermissionsRequest {
  permission_ids: string[]
}

export interface AssignRolesRequest {
  role_ids: string[]
}

export interface AssignUsersRequest {
  user_ids: string[]
}

export const roleApi = {
  // 获取角色列表
  getList: async (params: RoleListParams) => {
    const response = await client.get<Role[]>('/roles', { params })
    return response.data
  },

  // 获取单个角色
  getById: async (id: string) => {
    const response = await client.get<Role>(`/roles/${id}`)
    return response.data
  },

  // 创建角色
  create: async (data: RoleCreate) => {
    const response = await client.post<Role>('/roles', data)
    return response.data
  },

  // 更新角色
  update: async (id: string, data: RoleUpdate) => {
    const response = await client.put<Role>(`/roles/${id}`, data)
    return response.data
  },

  // 删除角色
  delete: async (id: string) => {
    await client.delete(`/roles/${id}`)
  },

  // 启用/停用角色
  updateStatus: async (id: string, status: boolean) => {
    const response = await client.patch<Role>(`/roles/${id}/status`, null, {
      params: { status },
    })
    return response.data
  },

  // 为角色分配权限
  assignPermissions: async (id: string, data: AssignPermissionsRequest) => {
    const response = await client.post<Role>(`/roles/${id}/permissions`, data)
    return response.data
  },

  // 获取角色的权限列表
  getPermissions: async (id: string) => {
    const response = await client.get<any[]>(`/roles/${id}/permissions`)
    return response.data
  },

  // 为用户分配角色
  assignRolesToUser: async (userId: string, data: AssignRolesRequest) => {
    const response = await client.post<any>(`/roles/users/${userId}/roles`, data)
    return response.data
  },

  // 为角色分配用户
  assignUsers: async (id: string, data: AssignUsersRequest) => {
    const response = await client.post<Role>(`/roles/${id}/users`, data)
    return response.data
  },

  // 获取角色的用户列表
  getUsers: async (id: string) => {
    const response = await client.get<any[]>(`/roles/${id}/users`)
    return response.data
  },
}

